File: problems/28-fenwick-tree/002-inversion-count-coordinate-compression/java/Solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class Solution {
    private static class FenwickTree {
        private final int[] tree;

        FenwickTree(int size) {
            tree = new int[size + 1];
        }

        void add(int index, int value) {
            while (index < tree.length) {
                tree[index] += value;
                index += index & -index;
            }
        }

        int sum(int index) {
            int result = 0;
            while (index > 0) {
                result += tree[index];
                index -= index & -index;
            }
            return result;
        }
    }

    public long countInversions(int[] nums) {
        if (nums.length < 2) {
            return 0;
        }

        int[] sorted = nums.clone();
        Arrays.sort(sorted);

        Map<Integer, Integer> rank = new HashMap<>();
        int nextRank = 1;
        for (int value : sorted) {
            if (!rank.containsKey(value)) {
                rank.put(value, nextRank++);
            }
        }

        FenwickTree bit = new FenwickTree(rank.size());
        long inversions = 0;
        int processed = 0;

        for (int value : nums) {
            int currentRank = rank.get(value);

            // Processed values greater than the current value.
            int lessOrEqual = bit.sum(currentRank);
            inversions += (long) processed - lessOrEqual;

            bit.add(currentRank, 1);
            processed++;
        }

        return inversions;
    }
}

File: problems/28-fenwick-tree/002-inversion-count-coordinate-compression/python/solution.py
class FenwickTree:
    def __init__(self, size: int):
        self.tree = [0] * (size + 1)

    def add(self, index: int, value: int) -> None:
        while index < len(self.tree):
            self.tree[index] += value
            index += index & -index

    def prefix_sum(self, index: int) -> int:
        result = 0
        while index > 0:
            result += self.tree[index]
            index -= index & -index
        return result


class Solution:
    def count_inversions(self, nums: list[int]) -> int:
        if len(nums) < 2:
            return 0

        sorted_values = sorted(set(nums))
        rank = {value: index for index, value in enumerate(sorted_values, start=1)}
        bit = FenwickTree(len(sorted_values))

        inversions = 0
        processed = 0

        for value in nums:
            current_rank = rank[value]

            # Earlier values greater than the current value.
            less_or_equal = bit.prefix_sum(current_rank)
            inversions += processed - less_or_equal

            bit.add(current_rank, 1)
            processed += 1

        return inversions
